File: backend/agents/ingredient_extractor_agent.py
"""Ingredient extraction and cleanup for nutrition pipeline."""


class IngredientExtractorAgent:
    ALLOWED_UNITS = {"g", "ml", "unit"}

    def extract_ingredients(self, recipe: dict) -> list[dict]:
        ingredients = recipe.get("ingredients", [])
        clean: list[dict] = []

        for ingredient in ingredients:
            name = str(ingredient.get("name", "")).strip().lower()
            amount = float(ingredient.get("amount", 0))
            unit = str(ingredient.get("unit", "")).strip().lower()

            if not name:
                raise ValueError("Ingredient is missing a name.")
            if amount <= 0:
                raise ValueError(f"Ingredient '{name}' has invalid amount.")
            if unit not in self.ALLOWED_UNITS:
                raise ValueError(f"Ingredient '{name}' has invalid unit '{unit}'.")

            clean.append({"name": name, "amount": amount, "unit": unit})

        return clean

File: supabase/functions/api/agents/ingredientExtractorAgent.ts
export function extractIngredients(recipe: Record<string, unknown>): Array<{ name: string; amount: number; unit: string }> {
  const rows = Array.isArray(recipe.ingredients) ? recipe.ingredients : [];
  return rows.map((r) => ({
    name: String((r as Record<string, unknown>).name || "").trim().toLowerCase(),
    amount: Number((r as Record<string, unknown>).amount || 0),
    unit: String((r as Record<string, unknown>).unit || "").trim().toLowerCase(),
  }));
}
